assembl/auth/make_saml.py

```diff
@@ -45,64 +45,109 @@ def private_key_from_cleaned_text(key_text):
-                   email='', alt_names=None, days=365):
+def make_saml_cert(key, country=None, state=None, locality=None, org=None,
+                   domain=None, email=None, alt_names=None, days=365,
+                   self_sign=True):
     if isinstance(key, basestring):
         key = private_key_from_cleaned_text(key)
     if alt_names and isinstance(alt_names, basestring):
         alt_names = alt_names.decode('utf-8').split()
-    subject = x509.Name([
-        # Provide various details about who we are.
-        x509.NameAttribute(NameOID.COUNTRY_NAME, country.decode('utf-8')),
-        x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, state.decode('utf-8')),
-        x509.NameAttribute(NameOID.LOCALITY_NAME, locality.decode('utf-8')),
-        x509.NameAttribute(NameOID.ORGANIZATION_NAME, org.decode('utf-8')),
-        x509.NameAttribute(NameOID.COMMON_NAME, cn.decode('utf-8')),
-        x509.NameAttribute(NameOID.EMAIL_ADDRESS, email.decode('utf-8')),
-    ])
+    name_components = []
+    if country:
+        name_components.append(x509.NameAttribute(
+            NameOID.COUNTRY_NAME, country.decode('utf-8')))
+    if state:
+        name_components.append(x509.NameAttribute(
+            NameOID.STATE_OR_PROVINCE_NAME, state.decode('utf-8')))
+    if locality:
+        name_components.append(x509.NameAttribute(
+            NameOID.LOCALITY_NAME, locality.decode('utf-8')))
+    if org:
+        name_components.append(x509.NameAttribute(
+            NameOID.ORGANIZATION_NAME, org.decode('utf-8')))
+    if domain:
+        name_components.append(x509.NameAttribute(
+            NameOID.COMMON_NAME, domain.decode('utf-8')))
+    if email:
+        name_components.append(x509.NameAttribute(
+            NameOID.EMAIL_ADDRESS, email.decode('utf-8')))
+    subject = x509.Name(name_components)
     pkey = key.public_key()
     skid = x509.SubjectKeyIdentifier.from_public_key(pkey)
-    # Create self-signed
-    serial_number = random_serial_number()
-    crt = x509.CertificateBuilder().subject_name(
+    if self_sign:
+        # Create self-signed
+        serial_number = random_serial_number()
+        builder = x509.CertificateBuilder()
+    else:
+        builder = x509.CertificateSigningRequestBuilder()
+    builder = builder.subject_name(
         subject
-    ).issuer_name(
-        subject
-    ).public_key(
-        pkey
-    ).serial_number(
-        serial_number
-    ).not_valid_before(
-        datetime.datetime.utcnow()
-    ).not_valid_after(
-        # Our certificate will be valid for 10 days
-        datetime.datetime.utcnow() + datetime.timedelta(days=days)
-    ).add_extension(
-        x509.BasicConstraints(True, 0), False
     ).add_extension(
-        skid, False
-    ).add_extension(
-        x509.AuthorityKeyIdentifier(
-            skid.digest, [x509.DirectoryName(subject)], serial_number), False
+        x509.KeyUsage(True, False, False, False, True, False, False, False, False), False
```

Comment:
Perhaps add the keynames here for readability? This is daunting to look at if you're not the writer.